docs(EXC-1744): Document canister backtraces (#5185)
dfinity/portal: docs/developer-docs/smart-contracts/maintain/backtraces.mdx

---
keywords: [intermediate, tutorial, maintain canisters, logs, logging, backtraces, debugging, trap, canister trap]
---

import { MarkdownChipRow } from "/src/components/Chip/MarkdownChipRow";
import { GlossaryTooltip } from "/src/components/Tooltip/GlossaryTooltip";

# Canister backtraces

<MarkdownChipRow labels={["Intermediate", "Tutorial", "Motoko", "Rust"]} />

## Intro

> Are you developing a Rust or Motoko <GlossaryTooltip>canister</GlossaryTooltip> encountering a confusing trap? Backtraces may help you pinpoint the location of the error. When canisters trap, ICP reports backtraces which contain the names of functions in the call stack. For example, if the following Rust canister accidentally performs an out of bounds access of stable memory:

Reviewer: ```suggestion
When canisters trap, ICP reports backtraces that contain the names of functions in the call stack. Backtraces are added to the canister logs and if the caller has [access to canister logs](/docs/current/developer-docs/smart-contracts/maintain/logs#log-visibility) they will also be included in the message error response.

For example, if the following Rust canister accidentally performs an out of bounds access of stable memory:
```

```
	#[update]
	fn outer() {
		inner();
	}

	fn inner() {
		inner_2();
	}

	fn inner_2() {
		ic_cdk::api::stable::stable_write(0xdeadbeef, b"foo")
	}
```

> then the resulting stable memory out of bounds error will also contain the following backtrace which can help determine that the error is in `inner_2` which is reached via `outer` and `inner`:

Reviewer: ```suggestion
The resulting stable memory out of bounds error will also contain the following backtrace which can help determine that the error is in `inner_2` which is reached via `outer` and `inner`:
```

```
Canister Backtrace:
ic0::ic0::stable64_write
_wasm_backtrace_canister::inner_2
_wasm_backtrace_canister::inner
_wasm_backtrace_canister::outer
```

> ## Where to find backtraces

Reviewer: ```suggestion

```

> Backtraces are added to the canister logs and if the caller has [access to canister logs](/docs/current/developer-docs/smart-contracts/maintain/logs#log-visibility) they will also be included in the message error response.

Reviewer: ```suggestion

```

Author: Why don't we want to include this information?
